pallets/attestation/src/migrations.rs

```diff
@@ -0,0 +1,133 @@
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use frame_support::{
+	traits::{Get, GetStorageVersion, OnRuntimeUpgrade, ReservableCurrency, StorageVersion},
+	weights::Weight,
+};
+use kilt_support::migration::switch_reserved_to_hold;
+use log;
+use sp_runtime::SaturatedConversion;
+use sp_std::marker::PhantomData;
+
+use crate::{AccountIdOf, Attestations, Config, CurrencyOf, HoldReason, Pallet};
+
+pub struct BalanceMigration<T>(PhantomData<T>);
+
+impl<T: crate::pallet::Config> OnRuntimeUpgrade for BalanceMigration<T>
+where
+	<T as Config>::Currency: ReservableCurrency<T::AccountId>,
+{
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		log::info!("Attestation: Initiating migration");
+
+		let onchain_storage_version = Pallet::<T>::on_chain_storage_version();
+		if onchain_storage_version.eq(&StorageVersion::new(1)) {
+			StorageVersion::new(2).put::<Pallet<T>>();
+			return do_migration::<T>();
+		}
+
+		log::info!(
+			"Attestation: No migration needed. This file should be deleted. Current storage version: {:?}, Required Version for update: {:?}", 
+			onchain_storage_version,
+			StorageVersion::new(1)
+		);
+
+		<T as frame_system::Config>::DbWeight::get().reads_writes(0, 0)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, &'static str> {
+		use sp_std::vec;
+
+		let has_all_user_no_holds = Attestations::<T>::iter_values()
+			.map(|details: crate::AttestationDetails<T>| {
+				kilt_support::migration::has_user_reserved_balance::<AccountIdOf<T>, CurrencyOf<T>>(
+					&details.deposit.owner,
+					&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				)
+			})
+			.all(|user| user);
+
+		assert!(
+			has_all_user_no_holds,
+			"Pre Upgrade Attestation: there are users with holds!"
+		);
+
+		assert_eq!(Pallet::<T>::on_chain_storage_version(), StorageVersion::new(1));
+
+		log::info!("Attestation: Pre migration checks successful");
+
+		Ok(vec![])
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(_pre_state: sp_std::vec::Vec<u8>) -> Result<(), &'static str> {
+		use frame_support::traits::fungible::InspectHold;
+
+		Attestations::<T>::iter().try_for_each(|(key, details)| -> Result<(), &'static str> {
+			let hold_balance: u128 = <T as Config>::Currency::balance_on_hold(
+				&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				&details.deposit.owner,
+			)
+			.saturated_into();
+			assert!(
+				details.deposit.amount.saturated_into::<u128>() <= hold_balance,
```

Comment:
I think we need to check the balance is exactly the same, right?

Reply:
I think it has to be `<=`, since one user can pay for multiple attestations which are composite into one hold.

Reply:
Does it make sense to sum up all deposits and all holds, and check for equality at the end?

Reply:
Yes it does and is implemeted in: [ce5debe08](https://github.com/KILTprotocol/kilt-node/pull/527/commits/ce5debe0817904bf699b3bd3b17cf6b060c6ece0)